docs/design/ConstructiveSolidGeometry.MD

```diff
@@ -0,0 +1,136 @@
+
+`m_PosSize.xyz` - position in 3D.
+
+`m_PosSize.w` - uniform scale. Non-uniform scaling can be added either
+
+`m_AuxData1` - auxiliary data for the command. What data is encoded here is specified by the `AuxCode1` code (`m_Control.y`)
+
+`m_AuxData2` - auxiliary data for the command. What data is encoded here is specified by the `AuxCode2` code (`m_Control.z`). Both auxiliary codes use the same list of codes and it's up to the user to decide which slot to use, but if both specify the same data it's undefined which one should be used (it's up to the implementation).
+
+`OpCode` (`m_Control.x`) - defines what the node does.
+
+`ControlCode` (`m_Control.w`) - control flags used by the shader.
+
+### Examples of some of the codes
+
+`OpCodeSphere` - the current node should trace a sphere using `sdSphere` and put the distance on the stack.
+
+`AuxCodeOpData` - the auxiliary data is operator data. Operator data is different per OpCode. For example, the `OpCodeSphere` requires a radius for the sphere it traces. If OpData is available, it gets it from GetOpData().x;
+
+`AuxCodeRotation` - the auxiliary data is rotation data for the node, given as a quaternion. See the Math section for details.
+
+## The Math
+
+Location: `//.../Shaders/Math.h`
+
+There are two important concepts in the math library - the use of quaternions for rotation, and the order of operations when traversing from parent node to child node.
+
+### Quaternions
+
+[Quaternions and spatial rotation](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation) - A spatial rotation of θ radians about a unit axis `(ux, uy, uz)` is mathematically given by `q = (cos(θ/2), ux * sin(θ/2), uy * sin(θ/2), uz * sin(θ/2))`. In the shader we choose to set the real component to `.w` and the imaginary components i, j and k to `.x`, `.y` and `.z` respectively . This means normalizing the  `.xyz` tuple gives you the rotation axis `(ux, uy, uz)`.
+
+To rotate a vector p (or a fixed point p) around the rotation axis, we need to evaluate the conjugation of p by q:
+```
+p(rotated) = q p q^(-1)
+```
+
+### Node inheritance
+
+Nodes define a position, an uniform scale and optionally a rotation quaternion. If they don't define a rotation, we use `(0, 0, 0, 1)` instead which is a 0 radian rotation.
+
+When we need to modify a point or a child node local to the current node, we 1. Scale, 2. Rotate and 3. Translate it. We can swap the scale and rotate operations because we use uniform scaling. Non-uniform scaling requires scale to be applied first.
+
+```
+new Pos = Rotate(Pos(child), Rot(parent)) * Scl(parent) + Pos(parent)
+new Rot = Rot(parent) * Rot(child)
+new Scl = Scl(parent) * Scl(child)
+```
+
+In the above assignment the order of multiplication for the rotation matters as it involves multiplication of their real and three imaginary components. Refer to the Wikipedia article for details.
```

Comment:
`above assignment the order` => `above assignment, the order`
`involves multiplication of their` => `involves the multiplication of their`